### ISSUE TITLE
Reconsider which heading levels should be available by default in rich text fields

### ISSUE THREAD

**Issue by thibaudcolas** [wagtail/wagtail#6535]
### Is your proposal related to a problem?

From the new [accessibility considerations](https://docs.wagtail.io/en/latest/advanced_topics/accessibility_considerations.html) documentation:

> Wagtail makes it very easy for developers to control which heading levels should be available for any given content, via :ref:`rich text features <rich_text_features>` or custom StreamField blocks.
In both cases, take the time to restrict what heading levels are available so the pages’ document outline is more likely to be logical and sequential. Consider using the following restrictions:
> 
> - Disallow ``h1`` in rich text. There should only be one ``h1`` tag per page, which generally maps to the page’s ``title``.
> - Limit heading levels to ``h2`` for the main content of a page. Add ``h3`` only if deemed necessary. Avoid other levels as a general rule.
> - For content that is displayed in a specific section of the page, limit heading levels to those directly below the section’s main heading.

This sparked some discussion [on Slack](https://wagtailcms.slack.com/archives/C014L7KJH3N/p1604945024071300) on which heading levels Wagtail makes available by default.

I think we should review those defaults, specifically to consider providing less by default, or even removing all heading levels from the default rich text formats.

### Describe the solution you'd like

Look at the set of default features and consider which really ought to be site-wide defaults for all Wagtail sites, and which would be better left to implementers to decide:

https://github.com/wagtail/wagtail/blob/be45cee2aa245debe36630813df43867592340e1/wagtail/core/wagtail_hooks.py#L41-L54

For heading levels in particular, the idea would be for implementers to make a conscious choice of which heading levels should be available in a given field, based on where the field is used, like they do for StreamField.

- Heading levels currently available in all fields by default: h2, h3, h4
- Could be reduced to either `h2, h3` only, or `h2`, or having no heading level available in rich text fields by default.

### Describe alternatives you've considered

The alternative would be to leave things as-is and let implementers decide for their site independently.

### Additional context

#6090, https://github.com/wagtail/wagtail/projects/10

**Comment by kaedroho** [comment]
Wonder if, instead of getting developers to define the specific heading levels they want. We instead always let editors use H1, H2 and H3 but let developers define a "shift" value that changes them at the point of rendering.

For example, if the developer chooses a shift of 2, this means H1 in the editor will become H3 when rendered and so on.

This has a couple of advantages:

 - It allows three distinct heading levels to be displayed in the editor. The editor doesn't need to care how they are actually rendered.
 - The developer can change the base shift level later and existing content wouldn't need to be rewritten.

**Comment by thibaudcolas** [comment]
This sounds to me like something that would be a better fit for StreamField than rich text? Having a generic "Heading" block that could then be rendered differently based on the context as you suggest. For rich text I think this would need the work suggested in #4223 to happen.

Here is an implementation of this concept in the React world: https://www.npmjs.com/package/react-accessible-headings, and [further reading](https://adrianroselli.com/2016/08/there-is-no-document-outline-algorithm.html) on why this hasn’t happened in vanilla HTML (yet?). 

For authors this will only make sense if headings in the editor don’t have a prescribed level attached to them, like in Dropbox Paper, so we’d need to change the UI a bit for this to work.

**Comment by thibaudcolas** [comment]
We discussed this with @helenb today. I don’t have precise notes but my recollection is:

- This feels like a worthwhile change, as part of which we could also encourage more usage of StreamField for heading levels so there is a clearer separation of concern between StreamField blocks and rich text. Not all Wagtail sites use StreamField though, so it’s not one-size-fits-all
- The most appealing options seem to be either removing `h4` and keeping `h2, h3`, or removing all heading levels altogether to encourage implementers to define this per field.
- There was some concern about this being a breaking change. If we do go ahead with this we’d need to be sure it’s worthwhile enough of a change.

If we do go ahead with any of this, here is a draft of what it would look like for sites who want to keep the current configuration – adding a new `wagtail_hooks.py` in one of their apps:

```py
from wagtail.core import hooks


@hooks.register('register_rich_text_features')
def register_core_features(features):
    # Wagtail 2.x removed h4, but we need it in all rich text fields – add it back.
    features.default_features.append('h4')
    # Wagtail 2.x removed heading levels, but we need them in all rich text fields – add them back.
    features.default_features.extend(['h2', 'h3', 'h4'])
    # --- Or alternatively,
    # Wagtail 2.x removed heading levels, but we need at least `h2` in all fields – add it back.
    features.default_features.append('h2')
```

**Comment by thibaudcolas** [comment]
Discussed in team meetings, the general sentiment is that this is too much potential annoyance for existing projects to justify the potential improvement. This may be more justified in a major Wagtail release (Wagtail 3.0 or otherwise).

For now I think I will pursue this by:

- Briefly reviewing code samples in the Wagtail docs for usage of `features` and [`WAGTAILADMIN_RICH_TEXT_EDITORS`](https://github.com/wagtail/wagtail/search?q=WAGTAILADMIN_RICH_TEXT_EDITORS)
- Coming up with a plan to support automated accessibility tests in Wagtail, that would catch heading issues. Either as part of the page edit’s form validation, or as part of preview (like Tota11y), or as part of workflows.

Also worth considering Karl’s RFC to replace the one-size-fits-all rich text fields with separate paragraph fields, which would also help with this: https://github.com/wagtail/rfcs/pull/46.
